File: JavaScript/1992-Find All Groups of Farmland.js
/**
 * @param {number[][]} land
 * @return {number[][]}
 */
var findFarmland = function(land) {
    let row = land.length, col = land[0].length;
    let res = new Array();
    
    let dfs = (r, c) =>{
        land[r][c] = 0;
        let maxr = r,  maxc = c;
        for (let [dr, dc] of [[1,0],[0,1],[-1,0],[0,-1]]){
            let nr = r+dr, nc = c+dc;
            if (0<=nr && nr<row && 0<=nc && nc<col && land[nr][nc] === 1){
                let [rr, cc] = dfs(nr, nc);
                maxr = Math.max(r, rr);
                maxc = Math.max(c, cc);
            }
        }
        return [maxr, maxc];
    }
    
    for (let r=0; r<row; r++){
        for (let c=0; c<col; c++){
            if (land[r][c] === 1){
                let [nr, nc] = dfs(r, c);
                res.push([r, c, nr, nc]);                
            }
        }
    }
    return res;
};

File: C++/1992-Find All Groups of Farmland.cpp
class Solution {
public:
    vector<vector<int>> findFarmland(vector<vector<int>>& land) {
        int row = land.size(), col = land[0].size();
        vector<vector<int>> res;
        
        for (int r = 0; r < row; r++) {
            for (int c = 0; c < col; c++) {
                if (land[r][c] == 1) {
                    int nr = r, nc = c;
                    for (nr=r; nr<row && land[nr][c]; nr++) {
                        for (nc=c; nc<col && land[nr][nc]; nc++) {
                            land[nr][nc] = 0;
                        }
                    }
                    res.push_back({r, c, nr-1, nc-1});
                }
            }
        }
        
        return res;
    }
};